### ISSUE TITLE
C++ API for torch.autograd.functional.jacobian

### ISSUE THREAD

**Issue by acxz** [pytorch/pytorch#40208]
## 🚀 Feature

Add C++ API for [`torch.autograd.functional.jacobian`](https://pytorch.org/docs/stable/_modules/torch/autograd/functional.html#jacobian)

## Motivation

I would like to calculate the jacobian of my model output wrt to the input and the feature is available in the python api but was dismayed to find out that it was not in the C++ API.

## Pitch

Include `torch::autograd::functional::jacobian` functionality in C++ API

## Alternatives

N/A

## Additional context

N/A

**Comment by vincentqb** [comment]
@albanD -- can you comment on this?

**Comment by albanD** [comment]
Yes the implementation is a bit tricky so the current version in only for python users.
We should add a corresponding API to cpp yes.

**Comment by Jaeyoung-Lim** [comment]
Any chance if there are any updates regarding this?

**Comment by albanD** [comment]
I'm afraid no one from the core team is working on that at the moment no. But we would accept a PR implementing this!

**Comment by swilson314** [comment]
I'm taking a stab at this, but wonder if I can get a hint about where to hook the code? 

Basically, rather than the gradient being accumulated, I want to grab the individual gradients (for my problem, one per case presented to the model) so I can populate the jacobian. I thought I found the appropriate place when I discovered autograd_meta->grad_accumulator_ and AccumulateGrad::apply(), but this only appears to get called once, *after* the gradient is accumulated -- i.e. a copy. I'm going to keep walking through the code....

**Comment by albanD** [comment]
Hi,

What you want to do is similar to the python version of these functions and use `autograd::grad()` that will return the gradients to you directly.

**Comment by swilson314** [comment]
Ah, I hadn't seen the vmap version of jacobian(), just what I'm looking for. Thanks!!

**Comment by swilson314** [comment]
It's likely I'm missing something, but I suspect the vectorize method won't be significantly faster on a CPU (vs GPU). 

What I don't understand is why it's not possible to get back a list/batch of tensors of the unaccumulated gradients. (This is certainly the model supported with forward(), with its ability to seamlessly handle batched  inputs/output.) I have a homegrown implementation of backprop where this isn't an issue, so I don't think it's a fundamental algorithm limitation.

**Comment by albanD** [comment]
> It's likely I'm missing something, but I suspect the vectorize method won't be significantly faster on a CPU (vs GPU).

You do reduce the number of backward calls you do as well as the number of framework calls so it does make a difference. Especially if your non-batched ops are small.

> list/batch of tensors of the unaccumulated gradients.

What do you mean by "unaccumulated gradient" here?

**Comment by swilson314** [comment]
In the code below, Experiments is a batch of training inputs. The result in Params is what I’m calling the accumulated gradient, but what I need to compute is the gradient per experiment.

	pTorchModel->zero_grad();
	auto Gradient = torch::ones(Experiments.sizes());
	Experiments.backward(Gradient);
	auto Params = pTorchModel->parameters();

The two methods below produce the desired gradient per experiment. Profiling shows both methods take almost the same time (this surprised me). This is orders of magnitude slower than our homegrown backprop() algorithm, which produces identical results.

	for (int i = 0; i < NCases; i++)
	{
		pTorchModel->zero_grad();
		Experiments[i].backward({}, true);
		GradientsToPtr(Params, &jac[(i*NVars)]);
	}

***

	auto Gradient = torch::zeros(Experiments.sizes());
	for (int i = 0; i < NCases; i++)
	{
		pTorchModel->zero_grad();
		Gradient[i] = 1.0; // Turn on gradient for just this case.
		Experiments.backward(Gradient, true);
		Gradient[i] = 0.0; // reset.
		GradientsToPtr(Params, &jac[(i*NVars)]);
	}


What I’m trying to create is something like this, where Result is a batched version of Params grads.

	torch::autograd::variable_list List;
	for (int i = 0; i < NCases; i++)
		List.push_back(Experiments[i]);
	auto Result = torch::autograd::grad(List, Params, /*IndividualGrads=*/true);
	for (int i = 0; i < NCases; i++)
		GradientsToPtr(Result[i], &jac[(i*NVars)]);

**Comment by swilson314** [comment]
By watching the graph execution, I have a guess that the info I want is created in the mat1_ix branch of AddmmBackward::apply().

Is this right?

If so, a couple hints would be greatly appreciated:
      how to best set a post_hook so I can examine the outputs of AddmmBackward?
      how to force AddmmBackward::apply() to compute the mat1_ix branch at the last variable?

**Comment by albanD** [comment]
From a mathematical point of view, what you want is not the gradient of the function you used in the forward.
It would be if the weights were expanded with an extra dimension corresponding to the number of cases.

If what you want are element-wise gradients, you can use libraries like https://github.com/cybertronai/autograd-hacks#per-example-gradients to use pytorch to compute "non-gradients" with the backward.

**Comment by swilson314** [comment]
Agreed, my example was intended to show what I meant by unaccumulated gradient. I actually need that, by output when there are multiple outputs. Thanks so much for the hint!

**Comment by swilson314** [comment]
@albanD I implemented a c++ version of the autograd-hacks code. Thanks again for pointing this out!

It currently only works for Linear and Conv2d per the code below. I suspect Conv1/3d are similar, but the einsum() is different? I hoped to find someplace in the libtorch gradient code that would give me  hint, but didn't have any luck. Do you happen to know how to extend these? Is there a more robust approach that uses already existing libtorch code for these operations so compute_grad1() works for the general case?

```
def compute_grad1(model: nn.Module, loss_type: str = 'mean') -> None:
    assert loss_type in ('sum', 'mean')
    for layer in model.modules():
        layer_type = _layer_type(layer)
        if layer_type not in _supported_layers:
            continue
        assert hasattr(layer, 'activations'), "No activations detected, run forward after add_hooks(model)"
        assert hasattr(layer, 'backprops_list'), "No backprops detected, run backward after add_hooks(model)"
        assert len(layer.backprops_list) == 1, "Multiple backprops detected, make sure to call clear_backprops(model)"

        A = layer.activations
        n = A.shape[0]
        if loss_type == 'mean':
            B = layer.backprops_list[0] * n
        else:  # loss_type == 'sum':
            B = layer.backprops_list[0]

        if layer_type == 'Linear':
            setattr(layer.weight, 'grad1', torch.einsum('ni,nj->nij', B, A))
            if layer.bias is not None:
                setattr(layer.bias, 'grad1', B)
 
        elif layer_type == 'Conv2d':
            A = torch.nn.functional.unfold(A, layer.kernel_size)
            B = B.reshape(n, -1, A.shape[-1])
            grad1 = torch.einsum('ijk,ilk->ijl', B, A)
            shape = [n] + list(layer.weight.shape)
            setattr(layer.weight, 'grad1', grad1.reshape(shape))
            if layer.bias is not None:
                setattr(layer.bias, 'grad1', torch.sum(B, dim=2))

```

**Comment by swilson314** [comment]
PS In my application, the new code is 560x faster, same as our home-grown backprop.

**Comment by swilson314** [comment]
Ok, I think I've got Conv1d and Conv3d figured out. Same einsum(), just needed to modify usage of unfold() since it only works for Conv2d.

It would be nicer if we were using the backend grad_fns rather than einsum(), but that would require more investigation.

Lastly, unless I'm missing something, I think this approach also works for multiple outputs, i.e. a "true" jacobian. Seems to work in the autograd-hacks code as long as you setup the gradients input to reflect the multiple outputs.

As I noted above, this is wildly faster than relying on repeated calls to backward() while varying the gradients input.

**Comment by swilson314** [comment]
I completed the c++ Jacobian (added Conv1d/3d) per my port of autograd-hacks. My approach is a kludge because I couldn't find a forward_hook mechanism for the C++ code like there is for python. As such, the model::forward() call needs covers around the forward calls made by Linear and ConvNd layers, plus a call to clear any existing hooks:

```
    virtual torch::Tensor forward(torch::Tensor& x) override
	{
		namespace F = torch::nn::functional;
		m_BackpropHooks.clear();
		// x = conv1(x);
		x = ForwardAndHook(torch::nn::AnyModule(conv1), x);
        	x = F::relu(x);
        	x = at::max_pool1d(x, 2, 2);
		x = ForwardAndHook(torch::nn::AnyModule(conv2), x);
        	x = F::relu(x);
        	x = at::max_pool1d(x, 2, 2);
		x = x.view({ -1, 4 * 50 });
		x = ForwardAndHook(torch::nn::AnyModule(fc1), x);
        	x = F::relu(x);
        	x = ForwardAndHook(torch::nn::AnyModule(fc2), x);
        	return x;
	}

```

It's less annoying for Sequential models:

```
	m_BackpropHooks.clear();
	auto it = m_Model->begin();
	x = ForwardAndHook(*it, x);
	for (++it; it != m_Model->end(); ++it)
		x = ForwardAndHook(*it, x);

```

I'm new to pytorch, so I'm not sure what to do next. It seems like adding a forward_hook for the C++ API would do the trick. This wouldn't be too bad for the Linear and ConvNd modules, but I'm not sure whether this could generalize.

**Comment by yaadhavraajagility** [comment]
Curious to know if there is a workaround for this
